rfcs/20190115-dense-attention.md

```diff
@@ -0,0 +1,506 @@
+  1. Reshape `query` and `value` into shapes `[batch_size, Tq, 1, dim]`
+     and `[batch_size, 1, Tv, dim]` respectively.
+  2. Calculate scores with shape `[batch_size, Tq, Tv]` as a non-linear
+     sum: `scores = tf.reduce_sum(tf.tanh(query + value), axis=-1)`
+  3. Use scores to calculate a distribution with shape
+     `[batch_size, Tq, Tv]`: `distribution = tf.nn.softmax(scores)`.
+  4. Use `distribution` to create a linear combination of `value` with
+     shape `batch_size, Tq, dim]`:
+     `return tf.matmul(distribution, value)`.
+
+  Args:
+    normalize: If True, will create scale and bias variables to normalize
+    scores.
+  """
+
+  def __init__(
+      self,
+      normalize=False,
+      **kwargs):
+
+  def build(self, input_shape):
+    """Creates variables."""
+
+  def call(self, inputs, mask=None):
+    """Applies additive attention.
+
+    Args:
+      inputs: List of the following tensors:
+        * query: Query `Tensor` of shape `[batch_size, Tq, dim]`.
+        * value: Value `Tensor` of shape `[batch_size, Tv, dim]`.
+      mask: List of the following tensors:
+        * query_mask: A boolean mask `Tensor` of shape `[batch_size, Tq]`.
+          If given, the output will be zero at the positions where
+          `mask==False`.
+        * value_mask: A boolean mask `Tensor` of shape `[batch_size, Tv]`.
+          If given, will apply the mask such that values at positions where
+          `mask==False` do not contribute to the result.
+    Returns:
+      Attention outputs of shape `[batch_size, Tq, dim]`.
+    """
+```
+
+The implementations for both Attention layers can be in the same file. They can
+reuse a private method with the following signature:
+
+```python
+def _apply_attention_scores(scores, value, value_mask=None):
+  """Applies attention scores to the given value tensor.
+
+  Args:
+    scores: Scores tensor of shape `[batch_size, Tq, Tv]`.
+    value: Value tensor of shape `[batch_size, Tv, dim]`.
+    value_mask: A boolean mask `Tensor` of shape `[batch_size, Tv]`.
+      If given, will apply the mask such that values at positions where
+      `mask==False` do not contribute to the result.
+
+  Returns:
+    Tensor of shape `[batch_size, Tq, dim]`.
+  """
+```
+
+Implementations of other Attention mechanisms can reuse this method, as well.
+So, that method can be made public. An alternative using inheritance is
+discussed in the "Base Attention Class" section. The
+rest of the code is specific to each Attention mechanism.
+
+Although not the primary focus of this proposal, the Attention layers work with
+RNN networks, such as
+[tf.keras.layers.LSTM](https://www.tensorflow.org/api_docs/python/tf/keras/layers/LSTM):
+When creating the LSTM, users need to set `return_sequences=True`, and the rest
+works the same way as CNN. It is unclear whether this method suffices to create
+the most common RNN+Attention models.
+
+We will first work on the implementation for Tensorflow.
+
+### Self-Attention
```

Comment:
Most uses of self-attention that I know of are seq2seq and so want masking.

Reply:
Can you clarify what you mean? I think the mask argument this api supports is good enough for DNN/CNN networks, such as transformer. WDYT?